### PR TITLE
Redesign Checkbox as a functional component and for Material theme

### PR DESCRIPTION
**Pull request by nicknisi**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

This pull request does the following:

- Converts Checkbox into a functional widget
- Replaces HyperScript calls with JSX
- Removes the toggle functionality from Checkbox (described in #796)
    - **Note:** This does not add a new toggle switch as a replacement for the dropped Checkbox functionality.
- Adds DOM required for Material theme (dojo/themes#68)

**Dojo Theme**

**Material Theme**

### CODE REVIEW

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-krzajvbnf.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-evibzxxyx.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-evibzxxyx.now.sh

File: src/checkbox/example/index.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { WidgetBase } from '@dojo/framework/core/WidgetBase';
```
Comment (tomdye) [inline src/checkbox/example/index.tsx]: can you remove this file and move it to the new example app format please?

File: src/checkbox/index.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+import { FocusProperties } from '@dojo/framework/core/mixins/Focus';
+import Label from '../label/index';
+import { formatAriaProperties } from '../common/util';
+import { uuid } from '@dojo/framework/core/util';
+
+export interface CheckboxProperties extends ThemedProperties, FocusProperties {
+	/** Custom aria attributes */
+	aria?: { [key: string]: string | null };
+	/**  Checked/unchecked property of the radio */
+	checked?: boolean;
```
Comment (tomdye) [inline src/checkbox/index.tsx]: can you add comments for the rest of these properties please?

File: src/checkbox/index.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+				secondary={true}
+			>
+				{label}
+			</Label>
+		)
+	];
+
+	return (
+		<div key="root" classes={getRootClasses()}>
+			{labelAfter ? children : children.reverse()}
```
Comment (tomdye) [inline src/checkbox/index.tsx]: can you ditch the label after part of this as we've done that elsewhere. Just place the label after the checkbox please?

File: src/checkbox/index.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+				forId={widgetId}
+				secondary={true}
+			>
+				{label}
+			</Label>
+		)
+	];
+
+	return (
+		<div key="root" classes={getRootClasses()}>
```
Comment (tomdye) [inline src/checkbox/index.tsx]: can you bring the `getRootClasses` functionality into the main render function please, no need for it to be a function

File: src/checkbox/index.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+				<svg classes={themeCss.checkmark} viewBox="0 0 24 24">
+					<path
+						classes={themeCss.checkmarkPath}
+						fill="none"
+						d="M1.73,12.91 8.1,19.28 22.79,4.59"
+					/>
+				</svg>
+				<div classes={themeCss.mixedMark} />
+			</div>
+			<div classes={themeCss.ripple} />
```
Comment (tomdye) [inline src/checkbox/index.tsx]: i really don't think we should be adding things like `ripple` classes / dom nodes

File: src/checkbox/index.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+			/>
+			<div classes={themeCss.background}>
+				<svg classes={themeCss.checkmark} viewBox="0 0 24 24">
+					<path
+						classes={themeCss.checkmarkPath}
+						fill="none"
+						d="M1.73,12.91 8.1,19.28 22.79,4.59"
+					/>
+				</svg>
+				<div classes={themeCss.mixedMark} />
```
Comment (tomdye) [inline src/checkbox/index.tsx]: what is mixedMark?

File: src/checkbox/index.tsx (hunk 5 above)
Comment (nicknisi) [reply src/checkbox/index.tsx]: @tomdye I don't like it either, particularly the svg element that's really only used for the material theme. Do you have a suggestion about how best to handle this?

File: src/checkbox/index.tsx (hunk 6 above)
Comment (nicknisi) [reply src/checkbox/index.tsx]: It's another class that's specifically used for the material theme. I could probably rename it to be something more generic, but it's not something that's really used by the current Dojo theme.

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-gj7wy2gxt.now.sh

File: src/checkbox/index.tsx (hunk 5 above)
Comment (tomdye) [reply src/checkbox/index.tsx]: we don't implement ripple elsewhere so i'd suggest you just leave it out

File: src/checkbox/index.tsx (hunk 6 above)
Comment (tomdye) [reply src/checkbox/index.tsx]: is it for a semi-selected state? I think it should just be removed to be honest

File: src/checkbox/index.tsx (hunk 5 above)
Comment (nicknisi) [reply src/checkbox/index.tsx]: I can remove it. Without the ripple there won't be a subtle hover effect on the checkbox.

**With ripple**
with-classes
There is a subtle hover highlight that appears

**Without ripple**
without-classes
The hover is no longer visible.

File: src/checkbox/index.tsx (hunk 6 above)
Comment (nicknisi) [reply src/checkbox/index.tsx]: Removed!

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* dojo.widgets: https://dojowidgets-78g8wuuvj.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-2huj2d0t1.now.sh
* Tests Deployment: https://dojowidgets-pay7ra89f.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-qhr5ti05m.now.sh
* Tests Deployment: https://dojowidgets-r6yjlnlkt.now.sh

File: src/checkbox/tests/unit/Checkbox.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+import * as sinon from 'sinon';
+import harness from '@dojo/framework/testing/harness';
+
+import Label from '../../../label/index';
+import Checkbox from '../../index';
+import * as css from '../../../theme/checkbox.m.css';
+import { noop, stubEvent } from '../../../common/tests/support/test-helpers';
+import focus from '@dojo/framework/core/middleware/focus';
+
+const compareId = {
```
Comment (tomdye) [inline src/checkbox/tests/unit/Checkbox.spec.tsx]: i think we have comparators for these in testHelpers already?

File: src/checkbox/tests/unit/Checkbox.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+} = {}) {
+	return () =>
+		middlewareFactory(() => ({
+			shouldFocus: () => shouldFocus,
+			focused: () => focused,
+			isFocused: () => isFocused
+		}))();
+}
+
+const expected = function(label = false, checked = false) {
```
Comment (tomdye) [inline src/checkbox/tests/unit/Checkbox.spec.tsx]: This should really be using an assertion template

Comment (nicknisi) [reply src/checkbox/tests/unit/Checkbox.spec.tsx]: Yeah, I just created a macro to convert what was there and didn't look too deeply at how the tests were structured. I went ahead and fixed that.

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-q4baxiibb.now.sh
* Tests Deployment: https://dojowidgets-2xl3mffoa.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-l6dn9qzx6.now.sh
* Tests Deployment: https://dojowidgets-13fiz4k0g.now.sh

**Pull request merged**